In [25]:
import pandas as pd

# train.csv dosyasını Pandas ile okuyoruz
df = pd.read_csv('../data/raw/train.csv')

# Verinin ilk 5 satırını ve sütunlarını inceliyoruz
df.head()

,Unnamed: 0,id,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,0,70172,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,...,5,4,3,4,4,5,5,25,18.0,neutral or dissatisfied
1,1,5047,Male,disloyal Customer,25,Business travel,Business,235,3,2,...,1,1,5,3,1,4,1,1,6.0,neutral or dissatisfied
2,2,110028,Female,Loyal Customer,26,Business travel,Business,1142,2,2,...,5,4,3,4,4,4,5,0,0.0,satisfied
3,3,24026,Female,Loyal Customer,25,Business travel,Business,562,2,5,...,2,2,5,3,1,4,2,11,9.0,neutral or dissatisfied
4,4,119299,Male,Loyal Customer,61,Business travel,Business,214,3,3,...,3,3,4,4,3,3,3,0,0.0,satisfied


In [ ]:
df.shape   # veri setinin satır ve sütun sayısını gösterir.

(103904, 25)

In [ ]:
df.info()  # veri setinin bilgilerini gösterir.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   id                                 103904 non-null  int64  
 2   Gender                             103904 non-null  object 
 3   Customer Type                      103904 non-null  object 
 4   Age                                103904 non-null  int64  
 5   Type of Travel                     103904 non-null  object 
 6   Class                              103904 non-null  object 
 7   Flight Distance                    103904 non-null  int64  
 8   Inflight wifi service              103904 non-null  int64  
 9   Departure/Arrival time convenient  103904 non-null  int64  
 10  Ease of Online booking             103904 non-null  int64  
 11  Gate location                      1039

In [ ]:
df.describe() # veri setinin istatistiksel özetini gösterir.

,Unnamed: 0,id,Age,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
count,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103594.000000
mean,51951.500000,64924.210502,39.379706,1189.448375,2.729683,3.060296,2.756901,2.976883,3.202129,3.250375,3.439396,3.358158,3.382363,3.351055,3.631833,3.304290,3.640428,3.286351,14.815618,15.178678
std,29994.645522,37463.812252,15.114964,997.147281,1.327829,1.525075,1.398929,1.277621,1.329533,1.349509,1.319088,1.332991,1.288354,1.315605,1.180903,1.265396,1.175663,1.312273,38.230901,38.698682
min,0.000000,1.000000,7.000000,31.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,25975.750000,32533.750000,27.000000,414.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,3.000000,3.000000,3.000000,2.000000,0.000000,0.000000
50%,51951.500000,64856.500000,40.000000,843.000000,3.000000,3.000000,3.000000,3.000000,3.000000,3.000000,4.000000,4.000000,4.000000,4.000000,4.000000,3.000000,4.000000,3.000000,0.000000,0.000000
75%,77927.250000,97368.250000,51.000000,1743.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,5.000000,4.000000,4.000000,4.000000,5.000000,4.000000,5.000000,4.000000,12.000000,13.000000
max,103903.000000,129880.000000,85.000000,4983.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,1592.000000,1584.000000


In [ ]:
df.describe(include="object") # veri setindeki kategorik değişkenlerin istatistiksel özetini gösterir.

,Gender,Customer Type,Type of Travel,Class,satisfaction
count,103904,103904,103904,103904,103904
unique,2,2,2,3,2
top,Female,Loyal Customer,Business travel,Business,neutral or dissatisfied
freq,52727,84923,71655,49665,58879


count -> sütunda boş olmayan kaç veri olduğunu gösterir.Tablo incelendiğinde bu sütunların hiçbirinde eksik (boş) veri olmadığı görülür.

unique -> O sütunun içinde kaç farklı seçenek olduğunu söyler.

top -> (En Popüler / Tepe Değer): O sütunda en çok tekrar eden cevabın ne olduğunu gösterir.

freq (Frekans / Tekrar Sayısı): Bir üst satırdaki o en popüler cevabın tam olarak kaç kez geçtiğini gösterir.

Özetle:
Müşterilerin çoğu kadın, çok büyük bir kısmı sadık müşteri, çoğunlukla iş seyahati (Business travel) yapıyorlar, en çok Business sınıfını tercih ediyorlar ama genel olarak memnun değiller.

In [ ]:
df.isna().sum() # veri setindeki eksik değerlerin sayısını gösterir.

Unnamed: 0                             0
id                                     0
Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Inflight wifi service                  0
Departure/Arrival time convenient      0
Ease of Online booking                 0
Gate location                          0
Food and drink                         0
Online boarding                        0
Seat comfort                           0
Inflight entertainment                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Inflight service                       0
Cleanliness                            0
Departure Delay in Minutes             0
Arrival Delay in Minutes             310
satisfaction    

Arrival Delay in Minutes sütununda 103594 non-null ama toplam satır 103904. Yani 310 satırda eksik değer var. Bunu veri temizleme adımında ele almamız gerekecek. (silmek mi, doldurmak mı örneğin ortalama/medyan ile karar vereceğiz.)
Diğer tüm sütunlarda eksik veri yok (hepsi 103904 non-null)

In [6]:
df.duplicated().sum() # veri setindeki tekrar eden satırların sayısını gösterir.

0

1. Memnuniyet Durumu

In [3]:
df['satisfaction'].value_counts()

satisfaction
neutral or dissatisfied    58879
satisfied                  45025
Name: count, dtype: int64

In [5]:
df['satisfaction'].value_counts(normalize=True) * 100

satisfaction
neutral or dissatisfied    56.666731
satisfied                  43.333269
Name: proportion, dtype: float64

Yolcuların yaklaşık %56.6'sı (58.879 kişi) uçuş deneyiminden "Nötr veya Memnun Değil" (neutral or dissatisfied).

Sadece %43.3'lük bir kesim (45.025 kişi) deneyiminden "Memnun" (satisfied) kalmış.

2. Müşteri Profili Sütunları

In [6]:
df['Customer Type'].value_counts()

Customer Type
Loyal Customer       84923
disloyal Customer    18981
Name: count, dtype: int64

In [7]:
df['Customer Type'].value_counts(normalize=True) * 100


Customer Type
Loyal Customer       81.732176
disloyal Customer    18.267824
Name: proportion, dtype: float64

### Müşteri Sadakati (Customer Type) Dağılımı

Havayolu şirketini tercih eden yolcuların sadakat profili incelendiğinde veri setinde belirgin bir yığılma görülmektedir.

Veri setindeki yolcuların **84.923'ü (%81.7)** Sadık Müşteriler (Loyal Customer), **18.981'i (%18.3)** ise Sadık Olmayan Müşteriler (Disloyal Customer) olarak sınıflandırılmıştır.

**İstatistiksel Çıkarım:**
Müşteri tabanının %80'inden fazlası sadık müşterilerden oluşmaktadır. Bu durum şirketin elindeki müşteriyi tutma konusunda oldukça güçlü bir konumda olduğunu gösterir. Ancak istatistiksel analiz ve makine öğrenmesi modellemesi aşamasında sınıflar arasındaki bu belirgin dengesizliği göz önünde bulundurmak kritik olacaktır. Sadık olmayan yolcuların azınlıkta olması, bu grubun memnuniyet/memnuniyetsizlik kalıplarını tespit etmeyi daha hassas bir süreç haline getirebilir.

In [8]:
df['Class'].value_counts()

Class
Business    49665
Eco         46745
Eco Plus     7494
Name: count, dtype: int64

In [9]:
df['Class'].value_counts(normalize=True) * 100

Class
Business    47.798930
Eco         44.988643
Eco Plus     7.212427
Name: proportion, dtype: float64

### Bilet Sınıfı (Class) Dağılımı

Havayolu şirketinin sunduğu bilet sınıflarına (kabin tiplerine) göre yolcu tercihleri incelendiğinde, ilk iki ana sınıf arasında oldukça yakın bir dağılım göze çarpmaktadır:

* **Business:** 49.665 yolcu (**%47.8**)
* **Eco (Ekonomi):** 46.745 yolcu (**%45.0**)
* **Eco Plus (Premium Ekonomi):** 7.494 yolcu (**%7.2**)

**İstatistiksel Çıkarım:**
Yolcu tabanının büyük çoğunluğu (%92.8) Business ve standart Ekonomi sınıfları arasında neredeyse yarı yarıya bölünmüştür. Eco Plus sınıfı ise müşteri kitlesi içinde oldukça küçük bir azınlığı temsil etmektedir. Makine öğrenmesi modellemesi aşamasında "Bilet Sınıfı" değişkeninin müşteri memnuniyeti üzerinde en belirleyici faktörlerden (feature) biri olma ihtimali yüksektir çünkü Business ve Eco uçuşları arasındaki fiziksel ve hizmetsel farklılıkların (koltuk konforu, yiyecek/içecek, diz mesafesi vb.) memnuniyet skorlarına doğrudan yansıması beklenir.

In [10]:
df['Type of Travel'].value_counts()

Type of Travel
Business travel    71655
Personal Travel    32249
Name: count, dtype: int64

In [11]:
df['Type of Travel'].value_counts(normalize=True) * 100

Type of Travel
Business travel    68.962696
Personal Travel    31.037304
Name: proportion, dtype: float64

### Seyahat Amacı (Type of Travel) Dağılımı

Havayolu şirketini tercih eden yolcuların uçuş amaçları incelendiğinde, iş seyahatlerinin belirgin bir ağırlığa sahip olduğu görülmektedir:

* **İş Seyahati (Business travel):** 71.655 yolcu (**%69.0**)
* **Kişisel Seyahat (Personal Travel):** 32.249 yolcu (**%31.0**)

**İstatistiksel Çıkarım:**
Yolcuların %69'unun uçuşlarını iş amaçlı gerçekleştirmesi, havayolu şirketinin kurumsal seyahat pazarında oldukça aktif olduğunu göstermektedir. Makine öğrenmesi modellemesi açısından bu değişken oldukça kritiktir çünkü iş seyahati yapan profesyonellerin hizmet öncelikleri (dakiklik, kesintisiz Wi-Fi, hızlı check-in) ile tatil veya kişisel seyahat yapan yolcuların öncelikleri genellikle birbirinden farklılık gösterir. Bu ayrım modelin yolcu memnuniyetini (satisfaction) tahmin etmesinde güçlü bir sinyal sağlayacaktır.

3. Anket (Servis Derecelendirme) Sütunları

In [12]:
df['Inflight wifi service'].value_counts()

Inflight wifi service
3    25868
2    25830
4    19794
1    17840
5    11469
0     3103
Name: count, dtype: int64

In [13]:
df['Inflight wifi service'].value_counts(normalize=True) * 100


Inflight wifi service
3    24.896058
2    24.859486
4    19.050277
1    17.169695
5    11.038074
0     2.986411
Name: proportion, dtype: float64

### Uçak İçi Wi-Fi Hizmeti (Inflight Wifi Service) Değerlendirmesi

Yolcuların uçuş esnasında sunulan Wi-Fi hizmetine verdikleri puanlar (0-5 ölçeğinde) incelendiğinde, memnuniyetin ağırlıklı olarak orta ve düşük seviyelerde (2 ve 3 puan) yoğunlaştığı görülmektedir:

* **3 Puan:** 25.868 yolcu (**%24.9**)
* **2 Puan:** 25.830 yolcu (**%24.9**)
* **4 Puan:** 19.794 yolcu (**%19.1**)
* **1 Puan (Çok Kötü):** 17.840 yolcu (**%17.2**)
* **5 Puan (Mükemmel):** 11.469 yolcu (**%11.0**)
* **0 (Kullanılmadı/Uygulanamaz):** 3.103 yolcu (**%3.0**)

**İstatistiksel Çıkarım:**
Veriler, Wi-Fi hizmetinin havayolu şirketi için belirgin bir "zayıf nokta" (pain point) olduğunu ortaya koymaktadır. Yolcuların neredeyse yarısı (%49.8) bu hizmete yalnızca 2 veya 3 puan vererek kötü bir deneyim yaşadıklarını belirtmiştir. Ayrıca, en düşük puanı (1) verenlerin oranının, en yüksek puanı (5) verenlerden çok daha yüksek olması dikkat çekicidir. Uçuşların büyük bir çoğunluğunun "İş Seyahati" olduğu (%69) düşünüldüğünde, internet altyapısındaki bu yetersizliğin genel yolcu memnuniyetsizliğini tetikleyen en önemli faktörlerden biri olması muhtemeldir.

4.Dijital Altyapı Kontrolü: 

Wi-Fi puanlarının çok düşük olduğunu bulmuştuk. Acaba bu havayolu şirketinin genel olarak "teknoloji ve dijitalleşme" konusunda bir problemi mi var? Yolcular uçağa binmeden önce web sitesi/uygulama üzerinden yapılan online biniş işlemlerine ne puan vermiş? Eğer bu da düşükse, sorunun kök nedeni şirketin teknolojik altyapısı olabilir.

In [17]:
df['Online boarding'].value_counts() 

Online boarding
4    30762
3    21804
5    20713
2    17505
1    10692
0     2428
Name: count, dtype: int64

In [18]:
df['Online boarding'].value_counts(normalize=True) * 100 


Online boarding
4    29.606175
3    20.984755
5    19.934747
2    16.847282
1    10.290268
0     2.336772
Name: proportion, dtype: float64

### Online Biniş (Online Boarding) İşlemi Değerlendirmesi

Yolcuların uçuş öncesi dijital ortamda gerçekleştirdikleri online biniş (check-in) deneyimlerine verdikleri puanlar incelendiğinde, görece dengeli ancak pozitif eğilimli bir dağılım dikkat çekmektedir:

* **4 Puan:** 30.762 yolcu (**%29.6**)
* **3 Puan:** 21.804 yolcu (**%21.0**)
* **5 Puan (Mükemmel):** 20.713 yolcu (**%19.9**)
* **2 Puan:** 17.505 yolcu (**%16.8**)
* **1 Puan (Çok Kötü):** 10.692 yolcu (**%10.3**)
* **0 (Kullanılmadı):** 2.428 yolcu (**%2.3**)

**İstatistiksel Çıkarım:**
Veriler, şirketin sunduğu online biniş hizmetinin genel olarak fena sayılmayacak bir seviyede olduğunu göstermektedir (yolcuların yaklaşık yarısı 4 ve 5 puan vermiştir). Ancak, müşteri profilinin büyük çoğunluğunun hız, pratiklik ve sorunsuz bir teknolojik deneyim bekleyen "İş Seyahati" (%69) yapan yolculardan oluştuğu düşünüldüğünde, en yüksek puan olan 5'in yalnızca %19.9'da kalması dikkat çekicidir. Bu durum, şirketin dijital check-in sürecinde (mobil uygulama veya web sitesi arayüzü) kullanıcı deneyimi (UX) açısından hala pürüzler olduğunu işaret etmektedir. Uçak içi Wi-Fi hizmetindeki bariz başarısızlık kadar acil bir kriz yaratmasa da, dijital altyapıya yapılacak iyileştirmelerin genel memnuniyet oranını yukarı çekme potansiyeli yüksektir.

5. Konfor ve Beklenti Eşleşmesi: Seat comfort

Yolcuların neredeyse %48'inin "Business" sınıfında uçtuğunu tespit edilmişti.
Eğer koltuk konforu puanları düşük veya orta seviyedeyse, bu durum şirket için ciddi bir alarmdır: "Müşterilere Business bilet satıyoruz ama onlara vaat ettiğimiz fiziksel konforu sağlayamıyoruz." 
Bu nedenle koltuk puanlarının çok yüksek (4 ve 5) olması beklenir.

In [19]:
df['Seat comfort'].value_counts() 

Seat comfort
4    31765
5    26470
3    18696
2    14897
1    12075
0        1
Name: count, dtype: int64

In [20]:
df['Seat comfort'].value_counts(normalize=True) * 100


Seat comfort
4    30.571489
5    25.475439
3    17.993532
2    14.337273
1    11.621304
0     0.000962
Name: proportion, dtype: float64

### Koltuk Konforu (Seat Comfort) Değerlendirmesi

Yolcuların uçuş esnasındaki koltuk konforuna verdikleri puanlar (0-5 ölçeğinde) incelendiğinde, değerlendirmelerin pozitif (yüksek puanlı) uçta yoğunlaştığı görülmektedir:

* **4 Puan:** 31.765 yolcu (**%30.6**)
* **5 Puan (Mükemmel):** 26.470 yolcu (**%25.5**)
* **3 Puan:** 18.696 yolcu (**%18.0**)
* **2 Puan:** 14.897 yolcu (**%14.3**)
* **1 Puan (Çok Kötü):** 12.075 yolcu (**%11.6**)
* **0 (Uygulanamaz/Boş):** 1 yolcu (**%0.001**)

**İstatistiksel Çıkarım:**
Veriler incelendiğinde, yolcuların yarısından fazlasının (%56.1) koltuk konforuna 4 ve 5 gibi yüksek puanlar verdiği tespit edilmiştir. Bu durum daha önceki analizlerimizde ortaya çıkan "yolcuların %47.8'inin Business sınıfını tercih ettiği" bulgusuyla istatistiksel açıdan güçlü bir tutarlılık göstermektedir. Şirketin donanımsal yatırımlarının (özellikle koltuk ergonomisi) müşteri tarafında karşılık bulduğu söylenebilir. Veri analizi ve modelleme aşamasında `Seat comfort` ile `Class` (Bilet Sınıfı) değişkenleri arasında yüksek bir korelasyon çıkması muhtemeldir. Ek olarak, "0" puan veren sadece 1 adet kaydın bulunması, istatistiksel analizlerde bir aykırı değer (outlier) veya veri girişi hatası olarak değerlendirilip incelenmeye adaydır.

6. Hizmet Kalitesi ve İnsan Faktörü: Inflight service veya On-board service

Wi-Fi gibi donanımsal sorunları gördük. Peki kabin memurlarının sunduğu hizmet nasıl? Eğer insan faktörüne dayalı hizmet puanları yüksekse, şirketin stratejisi "Donanımımız eksik ama personelimiz harika" şeklinde yorumlanabilir.

In [21]:
df['Inflight service'].value_counts()

Inflight service
4    37945
5    27116
3    20299
2    11457
1     7084
0        3
Name: count, dtype: int64

In [22]:
df['Inflight service'].value_counts(normalize=True) * 100


Inflight service
4    36.519287
5    26.097167
3    19.536303
2    11.026524
1     6.817832
0     0.002887
Name: proportion, dtype: float64

### Uçak İçi Hizmet (Inflight Service) Değerlendirmesi

Yolcuların uçuş esnasında aldıkları genel uçak içi hizmete (kabin memurlarının ilgisi, servis kalitesi vb.) verdikleri puanlar incelendiğinde, oldukça olumlu bir tablo çizildiği görülmektedir:

* **4 Puan:** 37.945 yolcu (**%36.5**)
* **5 Puan (Mükemmel):** 27.116 yolcu (**%26.1**)
* **3 Puan:** 20.299 yolcu (**%19.5**)
* **2 Puan:** 11.457 yolcu (**%11.0**)
* **1 Puan (Çok Kötü):** 7.084 yolcu (**%6.8**)
* **0 (Uygulanamaz/Boş):** 3 yolcu (**%0.0**)

**İstatistiksel Çıkarım:**
Uçak içi hizmet değerlendirmeleri, havayolu şirketinin en güçlü kaslarından birinin "insan faktörü ve servis kalitesi" olduğunu açıkça kanıtlamaktadır. Yolcuların büyük bir çoğunluğu (yaklaşık %62.6'sı) bu hizmete 4 ve 5 gibi yüksek puanlar vermiştir. Önceki analizlerde şirketin dijital altyapısında (özellikle Wi-Fi) ciddi zayıflıklar olduğunu tespit edilmişti ancak bu sonuçlar gösteriyor ki şirket, teknolojik eksikliklerini kabin ekibinin sunduğu başarılı insan odaklı hizmet ve yüksek koltuk konforu ile telafi etmektedir. Makine öğrenmesi modelimizin eğitimi aşamasında, `Inflight service` değişkeninin genel memnuniyeti (satisfaction) yukarı çeken en önemli pozitif özelliklerden  biri olarak öne çıkması beklenmektedir.

# DATA CLEANING 

In [26]:
df.drop(columns=["Unnamed: 0", "id"], inplace=True) # veri setindeki gereksiz sütunları kaldırır.

In [27]:
# 1. Tüm harfleri küçült
# 2. Boşlukları alt çizgi yap
# 3. Eğik çizgi (/) işaretlerini alt çizgi yap
# 4. Tire (-) işaretlerini alt çizgi yap
df.columns = (
    df.columns
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('/', '_')
    .str.replace('-', '_')
)

# İşlemin başarılı olup olmadığını görmek için sütun isimlerini yazdıralım:
print(df.columns)

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'class',
       'flight_distance', 'inflight_wifi_service',
       'departure_arrival_time_convenient', 'ease_of_online_booking',
       'gate_location', 'food_and_drink', 'online_boarding', 'seat_comfort',
       'inflight_entertainment', 'on_board_service', 'leg_room_service',
       'baggage_handling', 'checkin_service', 'inflight_service',
       'cleanliness', 'departure_delay_in_minutes', 'arrival_delay_in_minutes',
       'satisfaction'],
      dtype='object')


### Sütun İsimlerinin Standardizasyonu

Veri seti üzerinde yapılan incelemede, sütun isimlerinin büyük/küçük harf tutarsızlıkları, boşluklar ve özel karakterler ('/', '-') içerdiği görülmüştür (Örn: `Customer Type`, `Departure/Arrival time convenient`). 

Veri manipülasyonu ve makine öğrenmesi modelleme aşamalarında kod okunabilirliğini artırmak, sözdizimi (syntax) hatalarını önlemek ve metodolojik standartları sağlamak amacıyla tüm sütun isimleri **"snake_case"** formatına dönüştürülmüştür.

**Uygulanan Dönüşümler:**
* Tüm harfler küçük harf formatına (`lower()`) dönüştürüldü.
* İsimler arasındaki boşluklar, tire (`-`) ve eğik çizgi (`/`) işaretleri, alt çizgi (`_`) ile değiştirildi.

# Veri Tiplerinin Optimizasyonu

In [28]:
# Kategorik (metin içeren) sütunları seçiyoruz
kategorik_sutunlar = ['gender', 'customer_type', 'type_of_travel', 'class', 'satisfaction']

# Bir döngü ile hepsini tek tek 'category' tipine dönüştürüyoruz
for sutun in kategorik_sutunlar:
    df[sutun] = df[sutun].astype('category')

# Dönüşümün başarılı olduğunu ve hafıza tasarrufunu görmek için:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 23 columns):
 #   Column                             Non-Null Count   Dtype   
---  ------                             --------------   -----   
 0   gender                             103904 non-null  category
 1   customer_type                      103904 non-null  category
 2   age                                103904 non-null  int64   
 3   type_of_travel                     103904 non-null  category
 4   class                              103904 non-null  category
 5   flight_distance                    103904 non-null  int64   
 6   inflight_wifi_service              103904 non-null  int64   
 7   departure_arrival_time_convenient  103904 non-null  int64   
 8   ease_of_online_booking             103904 non-null  int64   
 9   gate_location                      103904 non-null  int64   
 10  food_and_drink                     103904 non-null  int64   
 11  online_boarding           

###  Veri Tipi Optimizasyonu (Memory Optimization)

Veri manipülasyonu aşamasında bellek (RAM) kullanımını minimize etmek ve ilerleyen fazlarda kurulacak makine öğrenmesi algoritmalarının eğitim hızını artırmak hedeflenmiştir. 

Bu doğrultuda, az sayıda farklı eşsiz değere sahip metinsel özellikler (`object` tipindeki sütunlar), istatistiksel modelleme prensiplerine uygun olarak `category` veri tipine dönüştürülmüştür. Bu işlem sayesinde veri setinin sistem üzerindeki yükü ciddi oranda hafifletilmiştir.

### Tekrarlayan (Kopya) Veri Kontrolü

In [29]:
# Önce veri setinde birbirinin tamamen aynısı kaç satır var onu bulalım
kopya_sayisi = df.duplicated().sum()
print(f"Tespit edilen kopya satır sayısı: {kopya_sayisi}")

# Eğer kopya satır varsa, bunları veri setinden kalıcı olarak siliyoruz
df.drop_duplicates(inplace=True)

# Temizlik sonrası elimizde kalan net satır sayısını kontrol edelim
print(f"Güncel ve temizlenmiş satır sayımız: {len(df)}")

Tespit edilen kopya satır sayısı: 0
Güncel ve temizlenmiş satır sayımız: 103904


### Tekrarlayan Verilerin (Duplicates) Kontrolü

Makine öğrenmesi modellerinde aşırı öğrenmeye (overfitting) ve algoritmik yanlılığa (bias) sebep olabilecek mükerrer kayıtlar `df.duplicated()` metodu ile incelenmiştir.

**Sonuç ve Çıkarım:**
Yapılan tarama sonucunda veri setinde **0 adet kopya satır** tespit edilmiştir. Mevcut 103.904 satırın tamamının eşsiz (unique) olduğu doğrulanmıştır. Veri toplama sürecinin oldukça temiz yürütüldüğünü ve veri bütünlüğünün kusursuz olduğunu gösteren bu sonuç sayesinde, herhangi bir satır silme işlemine gerek kalmadan bir sonraki analitik aşamaya geçilmiştir.